payjoin/src/hpke.rs

```diff
@@ -0,0 +1,283 @@
+            )),
+            &receiver_pk.0,
+            INFO_A,
+            &mut OsRng,
+        )?;
+    let aad = encapsulation_pair.public_key().to_bytes().to_vec();
+    let mut plaintext = reply_pk.to_bytes().to_vec();
+    plaintext.extend(body);
+    let plaintext = pad_plaintext(&mut plaintext, PADDED_PLAINTEXT_A_LENGTH)?;
+    let ciphertext = encryption_context.seal(plaintext, &aad)?;
+    let mut message_a = encapsulated_key.to_bytes().to_vec();
+    message_a.extend(&aad);
+    message_a.extend(&ciphertext);
+    Ok(message_a.to_vec())
+}
+
+#[cfg(feature = "receive")]
+pub fn decrypt_message_a(
+    message_a: &[u8],
+    receiver_sk: HpkeSecretKey,
+) -> Result<(Vec<u8>, HpkePublicKey), HpkeError> {
+    let enc = message_a.get(..UNCOMPRESSED_PUBLIC_KEY_SIZE).ok_or(HpkeError::PayloadTooShort)?;
+    let enc = EncappedKey::from_bytes(enc)?;
+    let aad = message_a
+        .get(UNCOMPRESSED_PUBLIC_KEY_SIZE..(UNCOMPRESSED_PUBLIC_KEY_SIZE * 2))
+        .ok_or(HpkeError::PayloadTooShort)?;
+    let encapsulation_pk = PublicKey::from_bytes(aad)?;
+    let mut decryption_ctx =
+        hpke::setup_receiver::<ChaCha20Poly1305, HkdfSha256, SecpK256HkdfSha256>(
+            &OpModeR::Auth(encapsulation_pk.clone()),
+            &receiver_sk.0,
+            &enc,
+            INFO_A,
+        )?;
+    let ciphertext = message_a.get(130..).ok_or(HpkeError::PayloadTooShort)?;
+    let plaintext = decryption_ctx.open(ciphertext, aad)?;
+    let reply_pk =
+        plaintext.get(..UNCOMPRESSED_PUBLIC_KEY_SIZE).ok_or(HpkeError::PayloadTooShort)?;
+    let reply_pk = HpkePublicKey(PublicKey::from_bytes(reply_pk)?);
+    let body = plaintext.get(UNCOMPRESSED_PUBLIC_KEY_SIZE..).ok_or(HpkeError::PayloadTooShort)?;
+    Ok((body.to_vec(), reply_pk))
+}
+
+/// Message B is sent from the receiver to the sender containing a Payjoin PSBT payload or an error
+#[cfg(feature = "receive")]
+pub fn encrypt_message_b(
+    mut plaintext: Vec<u8>,
+    receiver_keypair: &HpkeKeyPair,
+    sender_pk: &HpkePublicKey,
+) -> Result<Vec<u8>, HpkeError> {
+    let (encapsulated_key, mut encryption_context) =
+        hpke::setup_sender::<ChaCha20Poly1305, HkdfSha256, SecpK256HkdfSha256, _>(
+            &OpModeS::Auth((
+                receiver_keypair.secret_key().0.clone(),
+                receiver_keypair.public_key().0.clone(),
+            )),
+            &sender_pk.0,
+            INFO_B,
+            &mut OsRng,
+        )?;
+    let plaintext = pad_plaintext(&mut plaintext, PADDED_PLAINTEXT_B_LENGTH)?;
+    let ciphertext = encryption_context.seal(plaintext, &[])?;
+    let mut message_b = encapsulated_key.to_bytes().to_vec();
+    message_b.extend(&ciphertext);
+    Ok(message_b.to_vec())
+}
+
+#[cfg(feature = "send")]
+pub fn decrypt_message_b(
+    message_b: &[u8],
+    receiver_pk: HpkePublicKey,
+    sender_sk: HpkeSecretKey,
+) -> Result<Vec<u8>, HpkeError> {
+    let enc = message_b.get(..65).ok_or(HpkeError::PayloadTooShort)?;
```

Comment:
Can we replace the magic number (it seems like its (compressed_pk_size * 2 ) -1) or explain it in a comment?